nixos/lib/testing/matrix.nix

```diff
@@ -0,0 +1,255 @@
+          choices will be applied, and the depth of the relevant subtree
+          returned by `runTest` will be reduced by one node.
+
+          By avoiding useless test parameter combinations, you can reduce your testing workload,
+          and by avoiding nonsensical combinations, you can make all reasonable tests pass.
+
+          Make sure that `runTest` doesn't inspect this
+          decision too soon, thereby ensuring that all required definitions
+          are available. A typical use may look as follows:
+
+          ```nix
+          # For this example, we suppose that our application has a backend
+          # selection option, which only allowed when dynamic linking is enabled.
+
+          { linking, ... }:
+          {
+            matrix.backend = {
+              after = ["linking"];
+              enable = linking != "static";
+
+              # Neither choice module will be applied when static linking is enabled.
+              # If static linking does need some extra settings, set them in
+              # `matrix.linking.choice.static.module` instead.
+              choice.rest.module = ...;
+              choice.smtp.module = ...;
+            };
+          }
+          ```
+        '';
+      };
+    };
+  };
+
+  choiceModule = decisionName: extendModules: { name, ... }: {
+    options = {
+      module = mkOption {
+        description = mdDoc ''
+          The effects of making a choice. You can specify any test-level option here.
+
+          NixOS-level options can be specified in the {option}`nodes.<name>` and {option}`defaults` sub-options.
+        '';
+        example = { defaults.services.foo.backend = "xyz"; };
+        type = (extendModules { modules = [ (inMatrixModule decisionName name) ]; }).type;
+        default = { };
+        visible = "shallow";
+      };
+    };
+  };
+
+  inMatrixModule = decisionName: choiceName:
+    {
+      config = {
+        _module.args.${decisionName} = lib.mkDefault choiceName;
+        matrixDecisionsMade.${decisionName} = choiceName;
+        matrixIsRoot = false;
+      };
+    };
+
+in
+{
+  options = {
+    matrix = mkOption {
+      type = types.lazyAttrsOf (types.submodule decisionModule);
+      description = mdDoc ''
+        The `matrix` options let you define parameterized tests, test scenarios
+        with conditional configuration, test cases that run separately in isolated builds, and arbitrary combinations of those.
+
+        The commonality between these features is multiplicity, or equivalently from another perspective: choice.
+
+        This option constructs a hierarchy of options, where each decision to be made is represented by `<...>.matrix.<name>`.
+        These options themselves contain `choice.<name>`, which represent the possible values for each decision.
+
+        The `module` submodule is special, because it duplicates and extends the entire test configuration the existed before making the decision.
```

Comment:
```suggestion
        The `module` submodule is special, because it duplicates and extends the entire test configuration that existed before making the decision.
```